### PR TITLE
[ iOS Debug ] fast/loader/crash-replacing-location-before-load.html is a flaky crash

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### e11ee9df88e6ed81e827c30a3daa5103bf64292a
<pre>
[ iOS Debug ] fast/loader/crash-replacing-location-before-load.html is a flaky crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=251047">https://bugs.webkit.org/show_bug.cgi?id=251047</a>
rdar://104576815

Reviewed by Matt Woodrow.

New connections for, say remote rendering backend, are sent through
WP -&gt; GPUP main connection.
The new connection might be closed before the receiver receives it.
In this case, on Cocoa, the underlying connection Mach send right is
read from the kernel as MACH_PORT_DEAD during Mach message receipt.

Consider sequence:
 ... Runloop processing
 A
 Decode a message
 B
 Create connection from the port inside the message
 C
 ... Runloop processing

For IPC::Connection programming model, there is no difference
whether the sender closes the port at A, B or C. Make A behave
identical to already working B and C, e.g. deliver normal &quot;connection
did close&quot; signal.

Treat PORT_DEAD as a valid IPC::Connection client send right.
Construct the receive port as usual, but do not send the server
send right anywhere, as it cannot be sent to DEAD port. When the server
send right is destroyed, the receive port is notified that the sender
went away. This is obtains the common codepath for connection operation
for dead names.

* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformInvalidate):
(IPC::Connection::cancelSendSource):
(IPC::Connection::platformInitialize):
(IPC::Connection::platformOpen):
(IPC::Connection::sendOutgoingMessage):
(IPC::Connection::initializeSendSource):
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestMessageWithConnection::name):
(TestWebKitAPI::MockTestMessageWithConnection::arguments):
(TestWebKitAPI::MockTestMessageWithConnection::MockTestMessageWithConnection):
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/259582@main">https://commits.webkit.org/259582@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/0c29dd215cdeb4fb03a209599a0ab80c7404c5b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114508 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174696 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5249 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114443 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12006 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94968 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39472 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93853 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81144 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7663 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27961 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4561 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13809 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47516 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6605 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9546 "Built successfully") | | | 
| | | | |

### CODE REVIEW

**Comment by smfr** [comment]
> This is obtains

Doesn't parse

File: Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm [hunk 1]
```diff
@@ -154,41 +154,30 @@ static void clearNoSenderNotifications(mach_port_t port)
 
 void Connection::platformInitialize(Identifier identifier)
 {
-    if (!MACH_PORT_VALID(identifier.port))
-        return;
-
     if (m_isServer) {
+        ASSERT(MACH_PORT_VALID(identifier.port)); // Caller error. MACH_DEAD_NAME does not make sense, as we do not transfer receive rights.
```
Comment (smfr) [inline Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm]: Should this be RELEASE_ASSERT?

File: Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm [hunk 2]
```diff
@@ -154,41 +154,30 @@ static void clearNoSenderNotifications(mach_port_t port)
...
         m_receivePort = identifier.port;
-        m_sendPort = MACH_PORT_NULL;
-
 #if !PLATFORM(WATCHOS)
         mach_port_guard(mach_task_self(), m_receivePort, reinterpret_cast<mach_port_context_t>(this), true);
 #endif
     } else {
-        m_receivePort = MACH_PORT_NULL;
+        ASSERT(identifier.port != MACH_PORT_NULL);
```
Comment (smfr) [inline Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm]: Ditto

File: Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm [hunk 3]
```diff
@@ -346,16 +339,15 @@ static void clearNoSenderNotifications(mach_port_t port)
...
     return sendMessage(WTFMove(message));
 }
 
 void Connection::initializeSendSource()
 {
     ASSERT(m_isConnected);
-    ASSERT(MACH_PORT_VALID(m_sendPort));
+    if (m_sendPort == MACH_PORT_DEAD)
+        return;
+    ASSERT(m_sendPort != MACH_PORT_NULL);
```
Comment (smfr) [inline Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm]: Ditto

**Review by mattwoodrow — APPROVED** [review]

File: Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm [hunk 4]
```diff
@@ -205,17 +194,21 @@ static void clearNoSenderNotifications(mach_port_t port)
         m_isConnected = true;
 
         // Send the initialize message, which contains a send right for the server to use.
-        auto encoder = makeUniqueRef<Encoder>(MessageName::InitializeConnection, 0);
-
         mach_port_insert_right(mach_task_self(), m_receivePort, m_receivePort, MACH_MSG_TYPE_MAKE_SEND);
-        encoder.get() << MachSendRight::adopt(m_receivePort);
+        auto serverSendRight = MachSendRight::adopt(m_receivePort);
```
Comment (mattwoodrow) [inline Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:198]: This jump from native mach calls, into using our wrapper took a bit to wrap my head around (we're adopting just the newly allocated send right, but the port still has the existing receive right attached, so `m_receivePort` isn't invalid).

That's not new with your changes,  but it'd kinda be nice if we had `MachReceiveRight` which could then give us a newly allocated `MachSendRight`. One day :)

Comment (kkinnunen-apple) [reply Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:198]: If I ever understand the exact semantics of the port and port right namespaces, I'll implement it.

**Pull request merged**